File: android/app/src/main/java/com/example/barcodescanner/IrohBridge.kt
package com.example.barcodescanner

/**
 * JNI bridge to the Rust barcode-scanner library.
 *
 * All methods are blocking and must be called from a background thread.
 * The native library creates a tokio runtime internally.
 */
object IrohBridge {
    init {
        System.loadLibrary("barcode_scanner_android")
    }

    /**
     * Connects to a remote receiver using an EndpointTicket string.
     *
     * Returns an opaque session handle (non-zero on success, 0 on failure).
     */
    external fun connect(ticket: String): Long

    /**
     * Sends a scanned barcode/QR code to the connected receiver.
     *
     * [kind] is 0 for barcode, 1 for QR code.
     * [code] is the scanned string content.
     * [imageJpeg] is an optional JPEG image (can be empty).
     *
     * Returns true on success.
     */
    external fun sendScan(handle: Long, kind: Int, code: String, imageJpeg: ByteArray): Boolean

    /**
     * Returns true if the connection is still alive.
     */
    external fun isConnected(handle: Long): Boolean

    /**
     * Disconnects and frees the session handle.
     *
     * The handle must not be used after this call.
     */
    external fun disconnect(handle: Long)

    /**
     * Sends a sync poll to the receiver and returns the checked state for each code.
     *
     * [codesNl] is a newline-separated list of scanned code strings.
     *
     * Returns a newline-separated list of "code\u001fchecked" pairs where
     * checked is "1" (checked on receiver) or "0". Returns an empty string
     * if the poll fails or the connection is not available.
     */
    external fun syncCheckedState(handle: Long, codesNl: String): String
}

File: ios/BarcodeScanner/IrohBridge.swift
import Foundation

/// Swift wrapper around the Rust C-FFI barcode scanner library.
///
/// All blocking operations (connect, sendScan) are dispatched to a background
/// executor via `Task.detached` with `.background` priority so they don't block
/// the main actor.
///
/// The `handle` is an opaque pointer stored as `Int64` (same as `jlong` on Android).
/// It must be freed by calling `disconnect()`.
@MainActor
final class IrohBridge: ObservableObject {

    /// Non-zero when a live session exists.
    private var handle: Int64 = 0

    var isConnected: Bool {
        handle != 0 && barcode_scanner_is_connected(handle)
    }

    // MARK: – Connect

    /// Connects to the receiver at the given EndpointTicket string.
    ///
    /// Returns `true` on success.  Blocks the calling thread via the Rust
    /// Tokio runtime; **always call from a background context**.
    func connect(ticket: String) async -> Bool {
        let result: Int64 = await Task.detached(priority: .background) {
            ticket.withCString { ptr in
                barcode_scanner_connect(ptr)
            }
        }.value

        handle = result
        return result != 0
    }

    // MARK: – Send scan

    /// Sends a scanned code to the receiver.
    ///
    /// - Parameters:
    ///   - kind: 0 = barcode, 1 = QR code
    ///   - code: Raw scanned string
    ///   - imageJpeg: JPEG image data (may be empty)
    ///
    /// Returns `true` on success (scan ACKed within 15 seconds).
    func sendScan(kind: Int32, code: String, imageJpeg: Data) async -> Bool {
        let h = handle
        guard h != 0 else { return false }

        return await Task.detached(priority: .background) {
            code.withCString { codeCStr in
                if imageJpeg.isEmpty {
                    return barcode_scanner_send_scan(h, kind, codeCStr, nil, 0)
                } else {
                    return imageJpeg.withUnsafeBytes { rawBuf in
                        guard let basePtr = rawBuf.bindMemory(to: UInt8.self).baseAddress else {
                            return false
                        }
                        return barcode_scanner_send_scan(h, kind, codeCStr, basePtr, imageJpeg.count)
                    }
                }
            }
        }.value
    }

    // MARK: – Disconnect

    /// Closes the connection and frees the session handle.
    func disconnect() {
        let h = handle
        handle = 0
        if h != 0 {
            Task.detached(priority: .background) {
                barcode_scanner_disconnect(h)
            }
        }
    }
}
